File: frontend-food/src/schemas/ingredientReview.ts
import { z } from 'zod';

export const ReviewRecipeDraftSchema = z.object({
  title: z.string(),
  description: z.string().default(''),
  summary: z.string().default(''),
  servings: z.number().nullable(),
  preparation_time: z.number().nullable(),
  execution_time: z.number().nullable(),
  recipe_type: z.string(),
  difficulty: z.string().default('easy'),
  execution_time_choice: z.string().default('less_30'),
  preparation_time_choice: z.string().default('none'),
  scout_level_ids: z.array(z.number()).default([]),
  tag_ids: z.array(z.string()).default([]),
  steps: z.array(z.string()).default([]),
  source_url: z.string().default(''),
  image_url: z.string().default(''),
});

export const RecipeImportSourceSchema = z.object({
  type: z.enum(['url', 'text']),
  value: z.string().min(1),
});
export type RecipeImportSource = z.infer<typeof RecipeImportSourceSchema>;

export const ReviewSourceSchema = z.object({
  type: z.enum(['url', 'text']),
  label: z.string(),
  value: z.string(),
});
export type ReviewSource = z.infer<typeof ReviewSourceSchema>;

export const IngredientMatchCandidateSchema = z.object({
  id: z.number(),
  name: z.string(),
  slug: z.string().default(''),
  confidence: z.number().min(0).max(1),
});
export type IngredientMatchCandidate = z.infer<typeof IngredientMatchCandidateSchema>;

export const ReviewTechnicalDetailsSchema = z.object({
  method: z.string(),
  confidence: z.number().min(0).max(1),
  candidates: z.array(IngredientMatchCandidateSchema).default([]),
});

export const ReviewPortionSchema = z.object({
  id: z.number().nullable().default(null),
  name: z.string(),
  quantity: z.number().positive(),
  weight_g: z.number().positive().nullable().default(null),
  measuring_unit_id: z.number().nullable().default(null),
  measuring_unit_name: z.string().nullable().default(null),
  is_new: z.boolean().default(false),
});
export type ReviewPortion = z.infer<typeof ReviewPortionSchema>;

export const TemporaryIngredientDraftSchema = z.object({
  name: z.string(),
  description: z.string().default(''),
  status: z.string().default('draft'),
  values: z.record(z.unknown()).default({}),
  portions: z.array(ReviewPortionSchema).default([]),
  quantity: z.number().positive().nullable().default(null),
});
export type TemporaryIngredientDraft = z.infer<typeof TemporaryIngredientDraftSchema>;

export const IngredientReviewRowSchema = z.object({
  key: z.string(),
  source_text: z.string(),
  sources: z.array(ReviewSourceSchema).default([]),
  selected_ingredient_id: z.number().nullable().default(null),
  selected_ingredient_slug: z.string().default(''),
  selected_ingredient_name: z.string().default(''),
  suggested_ingredient_id: z.number().nullable().default(null),
  suggested_ingredient_name: z.string().default(''),
  candidates: z.array(IngredientMatchCandidateSchema).default([]),
  selected_portion: ReviewPortionSchema.nullable().default(null),
  suggested_portion: ReviewPortionSchema.nullable().default(null),
  quantity: z.number().positive().nullable().default(null),
  suggested_quantity: z.number().positive().nullable().default(null),
  reason: z.string().default(''),
  technical_details: ReviewTechnicalDetailsSchema.nullable().default(null),
  conflicts: z.array(z.string()).default([]),
  new_ingredient_draft: TemporaryIngredientDraftSchema.nullable().default(null),
  status: z.enum(['open', 'changed', 'confirmed', 'unresolved']).default('open'),
});
export type IngredientReviewRow = z.infer<typeof IngredientReviewRowSchema>;

export const IngredientReviewPreviewSchema = z.object({
  rows: z.array(IngredientReviewRowSchema).default([]),
  sources: z.array(ReviewSourceSchema).default([]),
  ai_interaction_id: z.string().nullable().default(null),
  recipe_draft: ReviewRecipeDraftSchema,
  // True when the page was unreachable and the data was reconstructed via
  // search grounding. The UI must ask the user to verify it.
  is_reconstructed: z.boolean().optional().default(false),
});
export type IngredientReviewPreview = z.infer<typeof IngredientReviewPreviewSchema>;

export const IngredientReviewRowInputSchema = z.object({
  key: z.string(),
  status: z.literal('confirmed'),
  selected_ingredient_id: z.number().nullable().default(null),
  selected_portion_id: z.number().nullable().default(null),
  quantity: z.number().positive(),
  temporary_ingredient: TemporaryIngredientDraftSchema.nullable().default(null),
});
export type IngredientReviewRowInput = z.infer<typeof IngredientReviewRowInputSchema>;

export const IngredientReviewFinalizeSchema = z.object({
  rows: z.array(IngredientReviewRowInputSchema).min(1),
});
export type IngredientReviewFinalize = z.infer<typeof IngredientReviewFinalizeSchema>;

export const ReviewFieldErrorSchema = z.object({
  path: z.string(),
  message: z.string(),
});

export const IngredientReviewErrorSchema = z.object({
  message: z.string(),
  fields: z.array(ReviewFieldErrorSchema).default([]),
});

File: backend/recipe/schemas/ingredient_review.py
"""Contracts for human review of AI-derived recipe ingredients."""

from typing import Any, Literal

from pydantic import BaseModel, Field

from .import_schemas import RecipeDraftOut

ReviewStatus = Literal["open", "changed", "confirmed", "unresolved"]
SourceType = Literal["url", "text"]


class RecipeImportSourceIn(BaseModel):
    """A source submitted for a combined recipe analysis."""

    type: SourceType
    value: str = Field(min_length=1, max_length=20_000)


class ReviewSourceOut(BaseModel):
    """Source reference retained for one extracted ingredient row."""

    type: SourceType
    label: str
    value: str


class IngredientMatchCandidateOut(BaseModel):
    id: int
    name: str
    slug: str = ""
    confidence: float = Field(ge=0, le=1)


class ReviewTechnicalDetailsOut(BaseModel):
    method: str
    confidence: float = Field(ge=0, le=1)
    candidates: list[IngredientMatchCandidateOut] = Field(default_factory=list)


class ReviewPortionOut(BaseModel):
    id: int | None = None
    name: str
    quantity: float = Field(gt=0)
    weight_g: float | None = Field(default=None, gt=0)
    measuring_unit_id: int | None = None
    measuring_unit_name: str | None = None
    is_new: bool = False


class TemporaryIngredientDraftOut(BaseModel):
    """Complete ingredient data shown in the existing ingredient editor."""

    name: str
    description: str = ""
    status: str = "draft"
    values: dict[str, Any] = Field(default_factory=dict)
    portions: list[ReviewPortionOut] = Field(default_factory=list)
    quantity: float | None = Field(default=None, gt=0, description="Suggested portion count from the import")


class IngredientReviewRowOut(BaseModel):
    """One immutable source row plus the current human-review proposal."""

    key: str
    source_text: str
    sources: list[ReviewSourceOut] = Field(default_factory=list)
    selected_ingredient_id: int | None = None
    selected_ingredient_slug: str = ""
    selected_ingredient_name: str = ""
    suggested_ingredient_id: int | None = None
    suggested_ingredient_name: str = ""
    candidates: list[IngredientMatchCandidateOut] = Field(default_factory=list)
    selected_portion: ReviewPortionOut | None = None
    suggested_portion: ReviewPortionOut | None = None
    quantity: float | None = Field(default=None, gt=0)
    suggested_quantity: float | None = Field(default=None, gt=0)
    reason: str = ""
    technical_details: ReviewTechnicalDetailsOut | None = None
    conflicts: list[str] = Field(default_factory=list)
    new_ingredient_draft: TemporaryIngredientDraftOut | None = None
    status: ReviewStatus = "open"


class IngredientReviewPreviewOut(BaseModel):
    """Common preview response for imports and existing-recipe AI actions."""

    rows: list[IngredientReviewRowOut] = Field(default_factory=list)
    sources: list[ReviewSourceOut] = Field(default_factory=list)
    ai_interaction_id: str | None = None
    recipe_draft: RecipeDraftOut
    # True when the page was unreachable and the data was reconstructed via
    # search grounding. The UI must ask the user to verify it.
    is_reconstructed: bool = False


class IngredientReviewRowIn(BaseModel):
    """Confirmed client-side state sent for final persistence."""

    key: str
    status: Literal["confirmed"]
    selected_ingredient_id: int | None = None
    selected_portion_id: int | None = None
    quantity: float = Field(gt=0)
    temporary_ingredient: TemporaryIngredientDraftOut | None = None


class IngredientReviewFinalizeIn(BaseModel):
    rows: list[IngredientReviewRowIn] = Field(min_length=1)


class ReviewFieldErrorOut(BaseModel):
    path: str
    message: str


class IngredientReviewErrorOut(BaseModel):
    message: str
    fields: list[ReviewFieldErrorOut] = Field(default_factory=list)
